File: scripts/decision-log-sync-check.mjs
#!/usr/bin/env node
// decision-log-sync-check.mjs
//
// Drift guard between CLAUDE.md "## 关键设计决策" body and docs/DECISION_LOG.md index table.
//
// Behavior:
//   1. Grep CLAUDE.md for `^N\. \*\*` (decision body anchors), find max N.
//   2. Count rows in DECISION_LOG.md table (lines matching `^| N |` where N is digits).
//   3. If max(CLAUDE.md) !== rowCount(DECISION_LOG.md), exit 1 with descriptive diff.
//   4. Also flag if any decision number in [1..maxN] is missing from EITHER side.
//
// Exit codes:
//   0 = in sync
//   1 = drift detected (PR block)
//   2 = file missing or unreadable

import { readFileSync, existsSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const REPO_ROOT = resolve(__dirname, '..');

const CLAUDE_MD = resolve(REPO_ROOT, 'CLAUDE.md');
const DECISION_LOG = resolve(REPO_ROOT, 'docs', 'DECISION_LOG.md');

function fail(msg, code = 1) {
  console.error(`[decision-log-sync-check] FAIL: ${msg}`);
  process.exit(code);
}

function ok(msg) {
  console.log(`[decision-log-sync-check] OK: ${msg}`);
}

if (!existsSync(CLAUDE_MD)) fail(`CLAUDE.md not found at ${CLAUDE_MD}`, 2);
if (!existsSync(DECISION_LOG)) fail(`DECISION_LOG.md not found at ${DECISION_LOG}`, 2);

const claudeText = readFileSync(CLAUDE_MD, 'utf8');
const logText = readFileSync(DECISION_LOG, 'utf8');

// Step 1 — find decision numbers in CLAUDE.md body
// Pattern: line starts with "N. **" where N is one or more digits
const claudeNumbers = new Set();
const claudeRe = /^(\d+)\. \*\*/gm;
let m;
while ((m = claudeRe.exec(claudeText)) !== null) {
  claudeNumbers.add(parseInt(m[1], 10));
}

if (claudeNumbers.size === 0) {
  fail('No decision anchors `^N\\. \\*\\*` found in CLAUDE.md — wrong file?');
}

const maxClaude = Math.max(...claudeNumbers);

// Step 2 — find decision numbers in DECISION_LOG.md index table
// Pattern: line starts with "| N |" where N is digits (the leading row column)
const logNumbers = new Set();
const logRe = /^\|\s*(\d+)\s*\|/gm;
while ((m = logRe.exec(logText)) !== null) {
  logNumbers.add(parseInt(m[1], 10));
}

if (logNumbers.size === 0) {
  fail('No table rows `| N |` found in DECISION_LOG.md — wrong file?');
}

const maxLog = Math.max(...logNumbers);

// Step 3 — compare max
if (maxClaude !== maxLog) {
  fail(
    `max decision number drift: CLAUDE.md = ${maxClaude}, DECISION_LOG.md = ${maxLog}. ` +
    `Either a decision body was added without an index row, or vice versa. Editing rule #1: same PR.`
  );
}

// Step 4 — gap check (every N in [1..max] must appear on both sides)
const missingFromClaude = [];
const missingFromLog = [];
for (let n = 1; n <= maxClaude; n++) {
  if (!claudeNumbers.has(n)) missingFromClaude.push(n);
  if (!logNumbers.has(n)) missingFromLog.push(n);
}

if (missingFromClaude.length > 0 || missingFromLog.length > 0) {
  const parts = [];
  if (missingFromClaude.length > 0) {
    parts.push(`missing from CLAUDE.md body: [${missingFromClaude.join(', ')}]`);
  }
  if (missingFromLog.length > 0) {
    parts.push(`missing from DECISION_LOG.md index: [${missingFromLog.join(', ')}]`);
  }
  fail(`gap detected — ${parts.join(' / ')}. Editing rule #3: numbers are monotonic, never renumber.`);
}

ok(`${maxClaude} decisions, fully in sync between CLAUDE.md body and DECISION_LOG.md index.`);
process.exit(0);

File: backend/scripts/decision_log_sync_check.py
"""Drift guard between CLAUDE.md body decisions and docs/DECISION_LOG.md index.

Plan J D4 (decision #25 Rule 6 / 2026-04-26): a CLAUDE.md decision body
without a matching DECISION_LOG.md index row (or vice versa) is the
canonical source of stale references. This guard enforces:

  1. The set of decision numbers in CLAUDE.md `^N\\. \\*\\*` body anchors
     equals the set of decision numbers in DECISION_LOG.md `^| N |`
     table rows.
  2. Numbers are monotonic with no gaps in `[1..max]` on either side.

Direct port of `scripts/decision-log-sync-check.mjs` (TS era, deleted in
Session A1' Step 8 .mjs sweep — decision #30.D / CLAUDE.md #29.B).
Behaviour and exit codes are preserved one-to-one.

Exit codes:
  0 — in sync
  1 — drift detected (PR block)
  2 — file missing or unreadable
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
CLAUDE_MD = REPO_ROOT / "CLAUDE.md"
DECISION_LOG = REPO_ROOT / "docs" / "DECISION_LOG.md"

_CLAUDE_RE = re.compile(r"^(\d+)\. \*\*", re.MULTILINE)
_LOG_RE = re.compile(r"^\|\s*(\d+)\s*\|", re.MULTILINE)


def _fail(msg: str, code: int = 1) -> int:
    print(f"[decision-log-sync-check] FAIL: {msg}", file=sys.stderr)
    return code


def _ok(msg: str) -> int:
    print(f"[decision-log-sync-check] OK: {msg}")
    return 0


def main() -> int:
    if not CLAUDE_MD.is_file():
        return _fail(f"CLAUDE.md not found at {CLAUDE_MD}", 2)
    if not DECISION_LOG.is_file():
        return _fail(f"DECISION_LOG.md not found at {DECISION_LOG}", 2)

    claude_text = CLAUDE_MD.read_text(encoding="utf-8")
    log_text = DECISION_LOG.read_text(encoding="utf-8")

    claude_numbers = {int(m.group(1)) for m in _CLAUDE_RE.finditer(claude_text)}
    if not claude_numbers:
        return _fail("No decision anchors `^N\\. \\*\\*` found in CLAUDE.md — wrong file?")

    log_numbers = {int(m.group(1)) for m in _LOG_RE.finditer(log_text)}
    if not log_numbers:
        return _fail("No table rows `| N |` found in DECISION_LOG.md — wrong file?")

    max_claude = max(claude_numbers)
    max_log = max(log_numbers)

    if max_claude != max_log:
        return _fail(
            f"max decision number drift: CLAUDE.md = {max_claude}, "
            f"DECISION_LOG.md = {max_log}. Either a decision body was added "
            "without an index row, or vice versa. Editing rule #1: same PR."
        )

    missing_from_claude = [n for n in range(1, max_claude + 1) if n not in claude_numbers]
    missing_from_log = [n for n in range(1, max_claude + 1) if n not in log_numbers]

    if missing_from_claude or missing_from_log:
        parts: list[str] = []
        if missing_from_claude:
            parts.append(f"missing from CLAUDE.md body: {missing_from_claude}")
        if missing_from_log:
            parts.append(f"missing from DECISION_LOG.md index: {missing_from_log}")
        return _fail(
            f"gap detected — {' / '.join(parts)}. Editing rule #3: numbers are "
            "monotonic, never renumber."
        )

    return _ok(
        f"{max_claude} decisions, fully in sync between CLAUDE.md body "
        "and DECISION_LOG.md index."
    )


if __name__ == "__main__":
    sys.exit(main())
